habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1289,16 +1402,553 @@ def step(
 
-        agent_observations = self._sim.get_observations_at(
-            position=final_position,
-            rotation=final_rotation,
+        # TODO: Make a better way to flag collisions
+        self._sim._prev_sim_obs["collided"] = collided  # type: ignore
+
+        # Update the state of the agent
+        self._sim.set_agent_state(  # type: ignore
+            final_position, final_rotation, reset_sensors=False
+        )
+
+        final_agent_state = self._sim.get_agent_state()
+        final_agent_state.position = final_position
+        final_agent_state.rotation = goal_rigid_state.rotation
+
+        return final_agent_state
+
+    def _apply_camera_pitch_velocity_action(
+        self,
+        camera_pitch_angular_velocity: float,
+        time_step: Optional[float] = None,
+    ):
+        """
+        Apply velocity command to the camera tilt angle for looking up and down
+        """
+        if time_step is None:
+            time_step = self._time_step
+
+        # Map velocity actions
+        self.vel_control.linear_velocity = mn.Vector3([0.0, 0.0, 0.0])
+        self.vel_control.angular_velocity = mn.Vector3(
+            [camera_pitch_angular_velocity, 0.0, 0.0]
+        )
+        sensor_names = list(self._sim.agents[0]._sensors.keys())  # type: ignore
+        sensor_state = self._sim.agents[0]._sensors[sensor_names[0]].node  # type: ignore
+
+        # Construct the sensor rigid state
+        agent_mn_quat = sensor_state.rotation
+        current_rigid_state = RigidState(
+            agent_mn_quat,
+            sensor_state.translation,
+        )
+
+        # manually integrate the rigid state
+        goal_rigid_state = self.vel_control.integrate_transform(
+            time_step, current_rigid_state
+        )
+
+        # Compute the delta increase
+        delta = np.sign(camera_pitch_angular_velocity) * abs(
+            float(goal_rigid_state.rotation.angle())
+            - float(current_rigid_state.rotation.angle())
+        )
+
+        # Get the camera pitch angle
+        camera_pitch_ang = self._get_camera_pitch_angle()
+
+        # Handle the min and max pitch angles
+        if camera_pitch_ang + delta > self._ang_range_camera_pitch[1]:
+            next_camera_pitch_ang = self._ang_range_camera_pitch[1]
+            goal_rigid_state.rotation = mn.Quaternion.rotation(
+                mn.Rad(next_camera_pitch_ang), mn.Vector3(1, 0, 0)
+            )
+        elif camera_pitch_ang + delta < self._ang_range_camera_pitch[0]:
+            next_camera_pitch_ang = self._ang_range_camera_pitch[0]
+            goal_rigid_state.rotation = mn.Quaternion.rotation(
+                mn.Rad(next_camera_pitch_ang), mn.Vector3(1, 0, 0)
+            )
+
+        # Update all the sensors
+        for sensor in sensor_names:
+            self._sim.agents[0]._sensors[  # type: ignore
+                sensor
+            ].node.rotation = goal_rigid_state.rotation
+
+        return self._sim.get_agent_state()
```

Comment:
Why is there a return on a `apply_` method ?

Reply:
This `_apply_velocity_action` method was originally within the `step` method of `VelocityAction`, implemented by @jimmytyyang.

I wanted to wrap all of this into a helper method because 1) the process of applying the velocity and integrating the sim is surprisingly complicated and having all of this in `step` will make the interface method unreadable; and 2) I wanted WaypointAction to also be able to use this. However I didn't have much context on how any of this worked so I just threw everything into a method, including the part where the next state of the agent was being queried at the end, which also seemed non-trivial.

Would love to know if there's a clean way to apply velocity so that we can directly query the next state. with `sim.get_agent_state()`